pkg/repo/repo.go

```diff
@@ -127,36 +150,264 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
-	targets := make([]TargetWithMetadata, 0, len(files))
+//
+// The targets will be added individually to the TUF repo if CreateRepoOptions.AddMetadataTargets
+// is set to true. The trusted_root.json file will be added if CreateRepoOptions.AddTrustedRoot
+// is set to true. At least one of these has to be true.
+func CreateRepoWithOptions(ctx context.Context, files map[string][]byte, options CreateRepoOptions) (tuf.LocalStore, string, error) {
+	if !options.AddMetadataTargets && !options.AddTrustedRoot {
+		return nil, "", errors.New("failed to create TUF repo: At least one of metadataTargets, trustedRoot must be true")
+	}
+
+	metadataTargets := make([]TargetWithMetadata, 0, len(files))
 	for name, bytes := range files {
-		var usage string
-		switch {
-		case strings.Contains(name, "fulcio"):
-			usage = "Fulcio"
-		case strings.Contains(name, "ctfe"):
-			usage = "CTFE"
-		case strings.Contains(name, "rekor"):
-			usage = "Rekor"
-		case strings.Contains(name, "tsa"):
-			usage = "TSA"
-		default:
-			usage = "Unknown"
-		}
-		scmActive, err := json.Marshal(&sigstoreCustomMetadata{Sigstore: CustomMetadata{Usage: usage, Status: "Active"}})
+		scmActive, err := json.Marshal(&sigstoreCustomMetadata{Sigstore: CustomMetadata{Usage: getTargetUsage(name), Status: "Active"}})
 		if err != nil {
 			return nil, "", fmt.Errorf("failed to marshal custom metadata for %s: %w", name, err)
 		}
-		targets = append(targets, TargetWithMetadata{
+		metadataTargets = append(metadataTargets, TargetWithMetadata{
 			Name:           name,
 			Bytes:          bytes,
 			CustomMetadata: scmActive,
 		})
 	}
 
+	targets := make([]TargetWithMetadata, 0, len(files)+1)
+	if options.AddMetadataTargets {
+		targets = append(targets, metadataTargets...)
+	}
+	if options.AddTrustedRoot {
+		trustedRootTarget, err := constructTrustedRoot(metadataTargets)
+		if err != nil {
+			return nil, "", fmt.Errorf("failed to construct trust root: %w", err)
+		}
+		targets = append(targets, *trustedRootTarget)
+	}
+
 	return CreateRepoWithMetadata(ctx, targets)
 }
 
+// CreateRepo calls CreateRepoWithOptions, while setting:
+// * CreateRepoOptions.AddMetadataTargets: true
+// * CreateRepoOptions.AddTrustedRoot: false
+func CreateRepo(ctx context.Context, files map[string][]byte) (tuf.LocalStore, string, error) {
+	return CreateRepoWithOptions(ctx, files, CreateRepoOptions{AddMetadataTargets: true, AddTrustedRoot: false})
+}
+
+func constructTrustedRoot(targets []TargetWithMetadata) (*TargetWithMetadata, error) {
+	var fulcioRoot, tsaLeaf, tsaRoot []byte
+	var fulcioIntermed, tsaIntermed [][]byte
+	rekorKeys := map[string]*root.TransparencyLog{}
+	ctlogKeys := map[string]*root.TransparencyLog{}
+	now := time.Now()
+
+	// we sort the targets by Name, this results in intermediary certs being sorted correctly,
+	// as long as there is less than 10, which is ok to assume for the purposes of this code
+	sort.Slice(targets, func(i, j int) bool {
```

Comment:
I think this is OK, but it does feel a bit fragile since it depends on file naming. Do we have any documentation in this repo we can point to to describe how we expect the names?

Reply:
What is setting the file names? I see this is passed in from https://github.com/sigstore/scaffolding/blob/main/cmd/tuf/server/main.go#L94-L126, are file names only controlled by Scaffolding or can they be set by users?

Reply:
I did this to handle how the [SplitCertChain](https://github.com/sigstore/scaffolding/blob/main/pkg/certs/certs.go#L35) function splits the TSA cert chain.

In terms of who sets the filenames - it's basically a list of files in a directory, so I guess the answer to that is "whoever creates that directory and puts files in there". I don't think we have any documentation about this. Where would such documentation go?

Reply:
Could we update https://github.com/sigstore/scaffolding?tab=readme-ov-file#tuf?